clr/commands.py

```diff
@@ -74,44 +76,158 @@ def resolve_command(query):
         print(f"Error! Command namespace '{namespace_key}' does not exist.\nClosest matches: {close_matches}\n\nAvailable namespaces: {sorted(NAMESPACE_KEYS)}", file=sys.stderr)
         sys.exit(1)
 
-    commands = list_commands(namespace_key)
-    if command_name not in commands:
-        close_matches = _get_close_matches(command_name, commands)
-        print(f"Error! Command '{command_name} does not exist in namespace '{namespace_key}'.\nClosest matches: {close_matches}\n\nAvailable commands: {commands}", file=sys.stderr)
+    namespace = cache.get(namespace_key) if cache else get_namespace(namespace_key)
+    if command_name not in namespace.commands:
+        close_matches = _get_close_matches(command_name, namespace.commands)
+        print(f"Error! Command '{command_name}' does not exist in namespace '{namespace_key}' - {namespace.descr}.\nClosest matches: {close_matches}\n\nAvailable commands: {namespace.commands}", file=sys.stderr)
         sys.exit(1)
 
-    return get_namespace(namespace_key), get_command(namespace_key, command_name), namespace_key, command_name
+    return namespace_key, command_name
 
-def get_command_spec(cmd):
+CommandSpec = namedtuple('CommandSpec', 'args varargs docstr')
+def get_command_spec(command_callable):
     """Get a command spec from the given (resolved) command, and
     distinguish default args vs. non-default args."""
-    args, vararg, varkwarg, defvals = inspect.getargspec(cmd)
 
-    assert varkwarg is None, 'Variable kwargs are not allowed in commands.'
+    # TODO(michael.cusack): Move to using Signature and remove deprecated
+    # getargspec.
+    args, vararg, varkwarg, defvals = inspect.getargspec(command_callable)
+    signature = Signature.from_callable(command_callable)
+
+    if signature.return_annotation != Signature.empty:
+        print(f'WARNING: {command_callable} returns a {signature.return_annotation} which is ignored.')
+    for param in signature.parameters.values():
+        if param.kind == param.VAR_KEYWORD:
+            print(f'WARNING: Ignoring kwargs found for clr command {param} {command_callable}: {varkwarg}')
 
     if args is None:
         args = tuple()
     if defvals is None:
         defvals = tuple()
 
     # Avoid the self argument.
-    if isinstance(cmd, types.MethodType):
+    if isinstance(command_callable, types.MethodType):
         args = args[1:]
 
     nargs = len(args) - len(defvals)
     args = list(zip(args[:nargs], [NO_DEFAULT]*nargs)) + list(zip(args[nargs:], defvals))
+    return CommandSpec(args, vararg, inspect.getdoc(command_callable))
+
+
+@dataclass
+class Namespace:
+    descr: str
+    longdescr: str
+    command_specs: dict
+    command_callables: dict
```

Comment:
If you wanted to be more specific, you could use the `typing` constructs, e.g. `Dict[str, Callable]`

Your call. :shrug: